File: backend/app/receipt/receipt.py
import os
from dateutil import parser
import hashlib
from app.cloud_vision.cloud_vision import CloudVision
from app.openai.openai import OpenAIClient
from app.receipt.utils import group_text_by_lines
from app.receipt.schema import Receipt, receipt_schema
from app.storage.storage import upload_image_bytes
import pytz

PROMPT = "Convert the given text of a receipt into a structured output format"

vision = CloudVision()
openai = OpenAIClient()


def hash_image_bytes(image_bytes: bytes) -> str:
    return hashlib.sha256(image_bytes).hexdigest()


class Extract:
    def __init__(self, image_bytes: bytes, fname: str):
        self.image_bytes = image_bytes
        self.fname = str
        self.image_hash = hash_image_bytes(image_bytes)
        self.file_extension = os.path.splitext(fname)[1][1:]

    def upload(self):
        obj_name = f"{self.image_hash}.{self.file_extension}"
        bucket = "receipts"
        upload_image_bytes(
            "receipts", obj_name, self.image_bytes, f"image/{self.file_extension}"
        )
        return bucket, obj_name

    def extract_text(self):
        response = vision.detect_text(self.image_bytes, self.image_hash)
        lines = group_text_by_lines(response.text_annotations)
        return "\n".join(lines)

    def structured_outut(self, input):
        return openai.text_to_json(PROMPT, input, "receipt_info", receipt_schema)

    def run(self):
        bucket, key = self.upload()
        text = self.extract_text()
        output = self.structured_outut(text)
        return self.to_model(output), text, bucket, key

    def to_model(self, output):
        dt = None
        date_str = output["opened"]
        try:
            dt = parser.parse(date_str)
            if dt.tzinfo is not None:
                dt = dt.astimezone(tz=None).replace(tzinfo=None)
        except Exception as e:
            print(f"[WARN] Unable to parse date string: {date_str}", e)
        output["opened"] = dt
        return Receipt(**output)

File: backend/internal/receipt/receipt.go
package receipt

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"log"
	"path/filepath"
	"strings"
	"time"

	"github.com/sharithg/civet/internal/cloudvision"
	"github.com/sharithg/civet/internal/genai"
	"github.com/sharithg/civet/internal/repository"
	"github.com/sharithg/civet/internal/storage"
)

const prompt = "Convert the given text of a receipt into a structured output format"

type Extract struct {
	ImageBytes   []byte
	FileName     string
	ImageHash    string
	FileExt      string
	visionClient *cloudvision.CloudVision
	openaiClient genai.OpenAi
	storage      storage.Storage
	Repo         *repository.Queries
}

func NewExtract(ctx context.Context, storage storage.Storage, openai genai.OpenAi, repo *repository.Queries, imageBytes []byte, fname string) (*Extract, error) {
	hash := sha256.Sum256(imageBytes)
	imageHash := hex.EncodeToString(hash[:])
	ext := strings.TrimPrefix(filepath.Ext(fname), ".")

	visionClient, err := cloudvision.NewCloudVision(ctx, "cache/cloud_vision")
	if err != nil {
		return nil, err
	}

	return &Extract{
		ImageBytes:   imageBytes,
		FileName:     fname,
		ImageHash:    imageHash,
		FileExt:      ext,
		visionClient: visionClient,
		openaiClient: openai,
		storage:      storage,
		Repo:         repo,
	}, nil
}

func (e *Extract) Upload(ctx context.Context) (bucket, key string, err error) {
	objectName := fmt.Sprintf("%s.%s", e.ImageHash, e.FileExt)
	bucket = "receipts"
	_, err = e.storage.UploadImageBytes(ctx, bucket, objectName, e.ImageBytes, "image/"+e.FileExt)
	return bucket, objectName, err
}

func (e *Extract) ExtractText(ctx context.Context) (string, error) {
	existing, err := e.Repo.GetCachedCloudVisionResponse(ctx, e.ImageHash)
	if err != nil {
		return "", err
	}

	if existing != nil {
		return strings.Join(existing, "\n"), nil
	}

	annotations, err := e.visionClient.DetectText(ctx, e.ImageBytes)
	if err != nil {
		return "", err
	}
	lines := GroupTextByLines(annotations, 10)
	_, err = e.Repo.InsertCachedCloudVisionResponse(ctx, repository.InsertCachedCloudVisionResponseParams{
		ImageHash: e.ImageHash,
		Response:  lines,
	})
	if err != nil {
		return "", err
	}

	return strings.Join(lines, "\n"), nil
}

func (e *Extract) StructuredOutput(ctx context.Context, input string) (Receipt, error) {
	var Schema = GenerateSchema[Receipt]()

	existing, err := e.Repo.GetCachedGenAiResponse(ctx, e.ImageHash)
	if err != nil {
		return Receipt{}, err
	}

	if existing != nil {
		var output Receipt
		err = json.Unmarshal(existing, &output)
		if err != nil {
			return Receipt{}, err
		}
		return output, nil
	}

	output, err := genai.JsonChat[Receipt](ctx, &e.openaiClient, prompt, input, "receipt_info", Schema)
	if err != nil {
		return Receipt{}, err
	}

	jsonOutput, err := json.Marshal(output)
	if err != nil {
		return Receipt{}, err
	}

	_, err = e.Repo.InsertCachedGenAiResponse(ctx, repository.InsertCachedGenAiResponseParams{
		ImageHash: e.ImageHash,
		Response:  jsonOutput,
	})
	if err != nil {
		return Receipt{}, err
	}

	return output, nil
}

func (e *Extract) Run(ctx context.Context) (ParsedReceipt, string, string, string, error) {
	bucket, key, err := e.Upload(ctx)
	if err != nil {
		return ParsedReceipt{}, "", "", "", err
	}

	text, err := e.ExtractText(ctx)
	if err != nil {
		return ParsedReceipt{}, "", "", "", err
	}

	out, err := e.StructuredOutput(ctx, text)
	if err != nil {
		return ParsedReceipt{}, "", "", "", err
	}

	model, err := e.ToModel(out)
	if err != nil {
		return ParsedReceipt{}, "", "", "", err
	}

	return model, text, bucket, key, nil
}

func (e *Extract) ToModel(output Receipt) (ParsedReceipt, error) {
	dateStr := output.Opened
	var parsed time.Time

	if dateStr != "" {
		t, err := time.Parse(time.RFC3339, dateStr)
		if err != nil {
			t, err = time.Parse("2006-01-02 15:04:05", dateStr)
		}
		if err != nil {
			log.Printf("[WARN] Unable to parse date string: %s, err: %v", dateStr, err)
		} else {
			parsed = t
		}
	}

	return ParsedReceipt{
		Receipt: Receipt{
			Restaurant:  output.Restaurant,
			Address:     output.Address,
			Opened:      output.Opened,
			OrderNumber: output.OrderNumber,
			OrderType:   output.OrderType,
			Table:       output.Table,
			Server:      output.Server,
			Items:       output.Items,
			Subtotal:    output.Subtotal,
			SalesTax:    output.SalesTax,
			Total:       output.Total,
			Payment:     output.Payment,
			Copy:        output.Copy,
			OtherFees:   output.OtherFees,
		},
		Opened: parsed,
	}, nil
}
